include/psa/crypto_values.h

```diff
@@ -2102,11 +2267,32 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
  *
  * The secret can also be a direct input (passed to
  * key_derivation_input_bytes()). In this case, the derivation operation
- * may not be used to derive keys: the operation will only allow
- * psa_key_derivation_output_bytes(), not psa_key_derivation_output_key().
+ * may not be used to derive or verify keys: the operation will only allow
+ * psa_key_derivation_output_bytes() or
+ * psa_key_derivation_verify_bytes() but not
+ * psa_key_derivation_output_key() or
+ * psa_key_derivation_verify_key().
  */
 #define PSA_KEY_DERIVATION_INPUT_SECRET     ((psa_key_derivation_step_t)0x0101)
 
+/** A low-entropy secret input for password hashing / key stretching.
+ *
+ * This is usually a key of type #PSA_KEY_TYPE_PASSWORD (passed to
+ * psa_key_derivation_input_key()) or a direct input (passed to
+ * psa_key_derivation_input_bytes()) that is a password or passphrase. It can
+ * also be high-entropy secret such as a key of type #PSA_KEY_TYPE_DERIVE or
+ * the shared secret resulting from a key agreement.
+ *
+ * The secret can also be a direct input (passed to
+ * key_derivation_input_bytes()). In this case, the derivation operation
+ * may not be used to derive or verify keys: the operation will only allow
+ * psa_key_derivation_output_bytes() or
+ * psa_key_derivation_verify_bytes(), not
+ * psa_key_derivation_output_key() or
+ * psa_key_derivation_verify_key().
```

Comment:
Thinking about it, I don't think it makes sense to prohibit `psa_key_derivation_verify_key` here. I expect it to be a common use case for password verification: the user has just entered a password, which the caller has in memory, so the input comes via `input_bytes` or requires the unnecessary overhead of a temporary key; the password hash is a non-exportable key object, so the output has to be consumed by `verify_key`.

This workflow is less common for non-password inputs, but I think `INPUT_SECRET` and `INPUT_PASSWORD` should have the same rules in this respect, so this change should apply to `INPUT_SECRET` as well.

Reply:
I see your point. Initially I thought there was a parallel with account creation (if the caller has the password in main memory, they also need the unnecessary overhead of a temporary key for the password in order to be able to use `output_key()`: but now I realise not so much: `output_key()` can easily be replaced with `output_bytes()` + `import_key()`, while the same doesn't hold for `verify_key()` (assuming the key is not exportable).

Reply:
The restriction on `output_key` is for application hygiene, not directly for security: if you're using some derived material as a key, and the material was derived from data that came from outside the crypto service, then the derived key itself comes from outside the crypto service. There's a desire (although I'm not sure how important this is) to allow implementations to keep track of which keys were always inside and which were potentially exposed, and this restriction very slightly simplifies `output_key`.

This hygiene restriction is on creating a key, so it only applies to `output_key`, not `verify_key`.

Reply:
Thanks, the reasoning behind the restriction on `ouput_key` was never entirely clear to me, which is probably one of the reasons I incorrectly transposed it to `verify_key`. It's much clearer now.

Reply:
The approach is not 'obviously sensible', as it leaves asymmetries in the process for enrolling a new password and verifying it later. Supposing that the input password is held in application memory, then enrolment looks like one of the following sequences:

1. Setup KDF operation for PBKDF
2. Input cost parameters with `input_number()`
3. Input salt with `input_bytes()`
4. Input password with `input_bytes()`
5. Erase password buffer
6. Extract password hash with `output_bytes()`
7. Store password hash with `import_key()` (type = PASSWORD_HASH, usage = VERIFY_DERIVATION, alg = PBKDF)
8. Erase hash buffer

Or by using a temporary key for the password:

1. Create temporary key with password (type = PASSWORD, usage = VERIFY_DERIVATION, alg = PBKDF)
2. Erase password buffer
3. Erase password buffer
4. Setup KDF operation for PBKDF
5. Input cost parameters with `input_number()`
6. Input salt with `input_bytes()`
7. Input password with `input_key()`
8. Store password hash with `output_key()` (type = PASSWORD_HASH, usage = VERIFY_DERIVATION, alg = PBKDF)
9. Destroy temp password key

In contrast, verification doesn't have the extra juggling step:

1. Setup KDF operation for PBKDF
2. Input cost parameters with `input_number()`
3. Input salt with `input_bytes()`
4. Input password with `input_bytes()`
5. Erase password buffer
6. Verify password hash with `verify_key()`

I think the flow for password verification has less application hygiene risks than those which are related to deriving keys - for the flows above I do not see why denying the simplification of directly storing the password hash is better?

Could the restriction on `output_key()` be triggered by the input function used for `INPUT_SECRET`, but not by `INPUT_PASSWORD` or `INPUT_SALT/PEPPER` etc? - which might deal with another of the open issues in this discussion?

Technically, we could infer what the `output_key()` will be used for based on its type or its usage policy, which could permit password-hash-storing usage but not cryptographic-key-derivation usage. This feels like it might be one level of complexity too far (for comprehension by either the users or implementers)?

Reply:
I think what you're suggesting makes sense in principle but I agree that the main question is that of complexity. As a general rule, I'm inclined towards making things simple to document even if it means the API a bit more restrictive that necessary, at least for a first version. For one thing, it's easier to lift restrictions in the future than to add them, for backwards compatibility reasons. Also as the API starts being used and we get feedback from users, we'll be in a better position to judge whether the additional flexibility is worth the additional complexity.

So, I'm inclined to keep the restriction on `ouput_key()` unchanged for now, and only lift the one on `verify_key()` as suggested by Gilles.